---
title: "Exercise 2 from R-for-Bioinformatics"
author: "Ramesh Chellaton"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown
#### Load libraries
```{r library & file load}
library(seqinr)
library(dplyr)
library(stringr)

dengue <- read.fasta('dengue.fasta')
dengue_seq <- dengue[[1]]
```
#### Q1: draw a sliding window plot of GC content
```{r sliding window plot of size 200 nucleotides}
GC_window_plot <- function(size, sequence) {
  x_values <- seq(1, length(sequence), by=size)
  y_values <- c()
  for (i in 1:length(x_values)) {
    GC_tmp <- GC(dengue_seq[i:(i+size)])
    y_values <- c(y_values, GC_tmp)
  }
  plot(x=x_values, y=y_values, type="b", col='steelblue2',
       main='GC plot', ylab='GC fraction')
}
GC_window_plot(200, dengue_seq)
```

#### Q2: GC plot for Mycobacterium with 20000 step
```{r leprae plot}
leprae <- read.fasta('leprae.fasta')
leprae_seq <- leprae[[1]]
GC_window_plot(20000, leprae_seq)
```

#### Q3: AT function - similar to GC in seqinr, except for 'A' and 'T'
```{r AT function}
AT <- function(seq) {
  monomer_tbl <- seqinr::count(seq, 1)
  res <- (sum(monomer_tbl['a']+monomer_tbl['t'])/length(seq))
  return(res)
}
print (AT(dengue_seq))
```
#### Q4: sliding window plot for AT
```{r generic function for a window slide 2-monomer fraction}
generic_window_plot <- function(fun, size, nseq, title='') {
  x_values <- seq(1, length(nseq), by=size)
  y_values <- c()
  for (i in 1:length(x_values)) {
    GC_tmp <- fun(nseq[i:(i+size)])
    y_values <- c(y_values, GC_tmp)
  }
  plot(x=x_values, y=y_values, type="b", col='steelblue2',
       main=title, 
       ylab=paste(as.character(substitute(fun)),'fraction'))
}
generic_window_plot(GC, 200, dengue_seq, 'GC Plot for dengue, step=200')
GC_window_plot(200, dengue_seq)
generic_window_plot(AT, 200, dengue_seq, title='AT Plot for dengue, step=200')
```

#### Q5: over/under-representation of 3-nucleotide GAC GC?
Use rho function that computes: f~xy~ / (f~x~ * f~y~) to get over-representation.
this generally works for dimers
```{r 3-nucleotide over/under-rep}

myrho <- function(sequence='', astr=''){
  if (is.null(sequence) | is.null(astr))
  {
    print("invalid input")
    return (0)
  }
  srch_tbl <- seqinr::count(sequence, wordsize = str_length(astr))
  monomer_tbl <- seqinr::count(sequence, wordsize = 1)
  srch_mono <- str_split(tolower(astr), '')
  srch_mono <- srch_mono[[1]]
  
  denom <- 1
  for (i in 1:length(srch_mono)) {
    denom <- denom * monomer_tbl[srch_mono[i]]/length(sequence)
  }
  numer <- srch_tbl[tolower(astr)]/length(sequence)
  return(numer/denom)
}

monomers <- seqinr::count(leprae_seq, 1)
dimers <- seqinr::count(leprae_seq, 2)
trimer <- seqinr::count(leprae_seq, wordsize=3)
rho_trimer <- rho(leprae_seq, wordsize = 3)
rho_trimer
print(paste('rho value for gac:', round(rho_trimer['gac'],8)))

print('using my rho function:')
res <- myrho(leprae_seq, 'gac')
print(res)
```
#### End of Exercise-2